demos/kdf/argon2.c

```diff
@@ -0,0 +1,151 @@
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include <openssl/core_names.h>
+#include <openssl/crypto.h>
+#include <openssl/kdf.h>
+#include <openssl/params.h>
+#include <openssl/thread.h>
+
+/*
+ * test vector from
+ * https://datatracker.ietf.org/doc/html/rfc9106
+ */
+
+/*
+ * Hard coding a password into an application is very bad.
+ * It is done here solely for educational purposes.
+ */
+static unsigned char password[] = {
+    0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
+    0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
+    0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01,
+    0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01, 0x01
+};
+
+/*
+ * The salt is better not being hard coded too.  Each password should have a
+ * different salt if possible.  The salt is not considered secret information
+ * and is safe to store with an encrypted password.
+ */
+static unsigned char salt[] = {
+    0x02, 0x02, 0x02, 0x02, 0x02, 0x02, 0x02, 0x02,
+    0x02, 0x02, 0x02, 0x02, 0x02, 0x02, 0x02, 0x02
+};
+
+/*
+ * Optional secret for KLDF
+ */
+static unsigned char secret[] = {
+    0x03, 0x03, 0x03, 0x03, 0x03, 0x03, 0x03, 0x03
+};
+
+/*
+ * Optional additional data
+ */
+static unsigned char ad[] = {
+    0x04, 0x04, 0x04, 0x04, 0x04, 0x04, 0x04, 0x04,
+    0x04, 0x04, 0x04, 0x04
+};
+
+/*
+ * Argon2 cost parameters
+ */
+static uint32_t memory_cost    = 32;
+static uint32_t iteration_cost =  3;
+static uint32_t parallel_cost  =  4;
+
+static const unsigned char expected_output[] = {
+   0x0d, 0x64, 0x0d, 0xf5, 0x8d, 0x78, 0x76, 0x6c,
+   0x08, 0xc0, 0x37, 0xa3, 0x4a, 0x8b, 0x53, 0xc9,
+   0xd0, 0x1e, 0xf0, 0x45, 0x2d, 0x75, 0xb6, 0x5e,
+   0xb5, 0x25, 0x20, 0xe9, 0x6b, 0x01, 0xe6, 0x59
+};
+
+int main(int argc, char **argv)
+{
+    int rv = 1;
+    EVP_KDF *kdf = NULL;
+    EVP_KDF_CTX *kctx = NULL;
+    unsigned char out[32];
+    OSSL_PARAM params[8], *p = params;
+    OSSL_LIB_CTX *library_context = NULL;
+    unsigned int threads;
+
+    library_context = OSSL_LIB_CTX_new();
+    if (library_context == NULL) {
+        fprintf(stderr, "OSSL_LIB_CTX_new() returned NULL\n");
+        goto end;
+    }
+
+    /* Fetch the key derivation function implementation */
+    kdf = EVP_KDF_fetch(library_context, "argon2id", NULL);
+    if (kdf == NULL) {
+        fprintf(stderr, "EVP_KDF_fetch() returned NULL\n");
+        goto end;
+    }
+
+    /* Create a context for the key derivation operation */
+    kctx = EVP_KDF_CTX_new(kdf);
+    if (kctx == NULL) {
+        fprintf(stderr, "EVP_KDF_CTX_new() returned NULL\n");
+        goto end;
+    }
+
+    /*
+     * Thread support can be turned off; use serialization if we cannot
+     * set requested number of threads.
+     */
+    threads = parallel_cost;
+    if (OSSL_set_max_threads(library_context, parallel_cost) != 1 ||
+        OSSL_get_max_threads(library_context) < parallel_cost)
```

Comment:
Would it be better to just cap threads at maximum available < parallel_cost?

Reply:
Ok, I added that...  but the logic was meant as if we create own library context, initially max_threads is always 0.
Setting max threads should then fail only if thread support is not compiled in (test vector uses 4 threads).